x-pack/plugins/encrypted_saved_objects/server/create_model_version.ts

```diff
@@ -0,0 +1,138 @@
+      throw new Error(
+        `An invalid Encrypted Saved Objects Model Version transformation is trying to transform across types ("${inputType.type}" => "${outputType.type}"), which isn't permitted`
+      );
+    }
+
+    const inputService = instantiateServiceWithLegacyType(inputType); // inputType
+    // ? instantiateServiceWithLegacyType(inputType)
+    // : encryptedSavedObjectsService;
+
+    const outputService =
+      inputType !== outputType ? instantiateServiceWithLegacyType(outputType) : inputService; // outputType
+    // ? instantiateServiceWithLegacyType(outputType)
+    // : encryptedSavedObjectsService;
+
+    const transformFn = createMergedTransformFn(
+      inputService,
+      outputService,
+      shouldTransformIfDecryptionFails,
+      incomingChanges
+    );
+
+    return { ...modelVersion, changes: [{ type: 'unsafe_transform', transformFn }] };
+  };
+
+function createMergedTransformFn(
+  inputService: Readonly<EncryptedSavedObjectsService>,
+  outputService: Readonly<EncryptedSavedObjectsService>,
+  shouldTransformIfDecryptionFails: boolean | undefined,
+  modelChanges: SavedObjectsModelChange[]
+): SavedObjectModelTransformationFn {
+  // This merges the functions from all 'data_backfill', 'data_removal', and 'unsafe_transform' changes
+  const mergedTransformFn = buildModelVersionTransformFn(modelChanges);
+
+  return (document, context) => {
+    const { type, id, originId } = document;
+
+    const descriptorNamespace = context.namespaceType === 'single' ? document.namespace : undefined;
+    const encryptionDescriptor = { id, type, namespace: descriptorNamespace };
+
+    // Note about isTypeBeingConverted: false
+    // "Converting to multi-namespace clashes with the ZDT requirement for serverless"
+    // See deprecation in packages/core/saved-objects/core-saved-objects-server/src/migration.ts SavedObjectMigrationContext
+    const documentToTransform = mapAttributes(document, (inputAttributes) => {
+      try {
+        return inputService.decryptAttributesSync<any>(encryptionDescriptor, inputAttributes, {
+          isTypeBeingConverted: false, // incompatible with model versions/ZDT
+          originId,
+        });
+      } catch (err) {
+        if (!shouldTransformIfDecryptionFails || !(err instanceof EncryptionError)) {
+          throw err;
+        }
+
+        context.log.warn(
+          `Decryption failed for encrypted Saved Object "${document.id}" of type "${document.type}" with error: ${err.message}. Encrypted attributes have been stripped from the original document and model version transformation will be applied but this may cause errors later on.`
+        );
+        return inputService.stripOrDecryptAttributesSync<any>(
+          encryptionDescriptor,
+          inputAttributes,
+          {
+            isTypeBeingConverted: false, // incompatible with model versions/ZDT
+            originId,
+          }
+        ).attributes;
+      }
+    });
```

Comment:
__optional nit:__ we construct identical decrypt params in two places, and also for every iteration. Maybe we can move them out of this callback to a parent function and re-use?
```suggestion
    const encryptionDescriptor = { id, type, namespace: descriptorNamespace };

    const decryptParams = {
      // "Converting to multi-namespace clashes with the ZDT requirement for serverless"
      // See deprecation in packages/core/saved-objects/core-saved-objects-server/src/migration.ts SavedObjectMigrationContext
      isTypeBeingConverted: false,
      originId,
    };

    const documentToTransform = mapAttributes(document, (inputAttributes) => {
      try {
        return inputService.decryptAttributesSync<any>(
          encryptionDescriptor,
          inputAttributes,
          decryptParams
        );
      } catch (err) {
        if (!shouldTransformIfDecryptionFails || !(err instanceof EncryptionError)) {
          throw err;
        }

        context.log.warn(
          `Decryption failed for encrypted Saved Object "${document.id}" of type "${document.type}" with error: ${err.message}. Encrypted attributes have been stripped from the original document and model version transformation will be applied but this may cause errors later on.`
        );
        return inputService.stripOrDecryptAttributesSync<any>(
          encryptionDescriptor,
          inputAttributes,
          decryptParams
        ).attributes;
      }
    });
```

Reply:
Great catch!